libs/runtime/ebpf_ring_buffer.c

```diff
@@ -228,27 +432,108 @@ ebpf_ring_buffer_map_buffer(_In_ const ebpf_ring_buffer_t* ring, _Outptr_ uint8_
     }
 }
 
+_Must_inspect_result_ _Ret_maybenull_ const ebpf_ring_buffer_record_t*
+ebpf_ring_buffer_next_consumer_record(
+    _Inout_ ebpf_ring_buffer_t* ring_buffer, _In_ const uint8_t* buffer, _Out_ size_t* end_offset)
+{
+    ebpf_ring_buffer_record_t* record = _ring_next_consumer_record(ring_buffer, end_offset);
+    if (record) {
+        return (ebpf_ring_buffer_record_t*)(buffer + ((uint8_t*)record - ring_buffer->shared_buffer));
+    } else {
+        return NULL;
+    }
+}
+
+#pragma warning(push)
+#pragma warning( \
+    disable : 28167) // warning C28167: Code analysis incorrectly reports that the function 'ebpf_ring_buffer_reserve'
+                     // does not restore the IRQL to the value that was current at function entry.
 _Must_inspect_result_ ebpf_result_t
 ebpf_ring_buffer_reserve(
     _Inout_ ebpf_ring_buffer_t* ring, _Outptr_result_bytebuffer_(length) uint8_t** data, size_t length)
 {
-    ebpf_result_t result;
-    ebpf_lock_state_t state = ebpf_lock_lock(&ring->lock);
-    ebpf_ring_buffer_record_t* record = _ring_buffer_acquire_record(ring, length);
-    if (record == NULL) {
-        result = EBPF_INVALID_ARGUMENT;
-        goto Done;
+    uint32_t record_header = (uint32_t)length | EBPF_RINGBUF_LOCK_BIT;
+    size_t ring_capacity = _ring_get_length(ring);
+    size_t record_size = _ring_record_size(length);
+    size_t padded_record_size = _ring_padded_size(record_size);
+    if (padded_record_size > ring_capacity || length == 0 || length > (1ULL << 30)) {
+        return EBPF_INVALID_ARGUMENT;
     }
 
-    record->header.locked = 1;
-    MemoryBarrier();
+    size_t consumer_offset = ReadULong64NoFence(&ring->consumer_offset);
+    size_t reserve_offset = ReadULong64Acquire(&ring->producer_reserve_offset);
+    size_t used_capacity = reserve_offset - consumer_offset;
+    if (used_capacity + padded_record_size > ring_capacity) {
+        return EBPF_NO_MEMORY;
+    }
 
-    *data = record->data;
-    result = EBPF_SUCCESS;
+    //  Reservation loop:
+    //  - No fairness guarantee, but does guarantee progress on each race/collision
+    //  - Still needs to run at dispatch (or if all threads were passive you could yield in the spin loop)
```

Comment:
Is it possible that we could end up in a scenario where we keep looping (not necessarily infinitely, but just for a long time)? I.e can we hit DPC watchdog timeout (basically the thread remaining at dispatch for 'too long')?